src/main/java/org/apache/sling/models/impl/AssignableFrom.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.sling.models.impl;
+
+public interface AssignableFrom {
```

Comment:
i'm wondering if it's worth to provide this as a generic helper class supporting all types of checks - but it's only used for checking the (unknown) record class. suggestions:
* refactor the helper class to only actually check for java.lang.Record - then it's not required to put "java.lang.Records" in multiple other places in the code
* probably it's not even required to use "isAssignableFrom", afaik it's not possible to extend a Record class in java language by design
* do not use a interface for this helper class - it's not used as an interface
* do not make it public (even if part of impl package)

Reply:
You are correct. There is no need to create a generic `AssignableFrom`. I have made the respective changes to the code. The mechanism of the current implementation is as follows:

1. Check if the object being constructed is a record. SLING-12359 pertains exclusively to records and no other types should be affected.
2. The first Java LTS that provides native API to check if a given type is a record is Java 17 (`java.lang.Class#isRecord()`). However, as far as I'm concerned, neither current nor the nearest future releases of Apache Sling Models Implementation will require Java 17 as the minimum version. Therefore, until Java 17 is adopted as the minimum version, the most reasonable way to check if a type of object being constructed is a record is via `java.lang.Class#isAssignableFrom(String)`.
3. When using `java.lang.Class#isAssignableFrom(String)` to check if a type of object being constructed is a record, `java.lang.Record` is passed as an argument. The reason for that is that according to the Java Language Specification all record types have `java.lang.Record` as their implicit and direct superclass: https://docs.oracle.com/javase/specs/jls/se21/html/jls-8.html#jls-8.1.4.
4. If the type of object being constructed is a record and it doesn't have any explicit constructor with `@Inject` like normal Sling Model classes do, the construction logic is adjusted accordingly. The adjustment is to use a record's implicit canonical constructor, because records always have a single implicit canonical constructor that always instantiates all fields of a given record: https://docs.oracle.com/javase/specs/jls/se21/html/jls-8.html#jls-8.10.4.

Please let me know if any further changes to the implementation logic are required. I'd be glad to introduce them.